File: pixie/instrumentation/instrumentors.py
"""Auto-discovers and activates OpenInference instrumentors."""

from __future__ import annotations

import importlib
import logging

logger = logging.getLogger("pixie.instrumentation")

_KNOWN_INSTRUMENTORS = [
    ("openinference.instrumentation.openai", "OpenAIInstrumentor"),
    ("openinference.instrumentation.anthropic", "AnthropicInstrumentor"),
    ("openinference.instrumentation.langchain", "LangChainInstrumentor"),
    ("openinference.instrumentation.google_genai", "GoogleGenAIInstrumentor"),
    ("openinference.instrumentation.dspy", "DSPyInstrumentor"),
    # OTel official OpenAI v2 as secondary fallback
    ("opentelemetry.instrumentation.openai_v2", "OpenAIInstrumentor"),
]


def _activate_instrumentors() -> list[str]:
    """Try to instrument all known LLM providers. Returns list of activated names."""
    activated: list[str] = []
    for module_path, class_name in _KNOWN_INSTRUMENTORS:
        try:
            module = importlib.import_module(module_path)
            cls = getattr(module, class_name)
            cls().instrument()
            activated.append(class_name)
            logger.debug("Activated instrumentor: %s from %s", class_name, module_path)
        except ImportError:
            logger.debug(
                "Instrumentor not available (package not installed): %s", module_path
            )
        except Exception:
            logger.debug(
                "Instrumentor failed to activate: %s from %s",
                class_name,
                module_path,
                exc_info=True,
            )
    if not activated:
        logger.warning(
            "No LLM instrumentors activated. Install provider packages "
            "(e.g. openinference-instrumentation-openai) for auto-capture."
        )
    return activated

File: ts/src/instrumentation/instrumentors.ts
/**
 * Auto-discovers and activates known LLM instrumentor packages.
 *
 * Tries to `require()` known OpenInference instrumentor packages
 * and activate them. Missing packages are silently skipped.
 */

const KNOWN_INSTRUMENTORS: Array<[string, string]> = [
  ["@arizeai/openinference-instrumentation-openai", "OpenAIInstrumentor"],
  [
    "@arizeai/openinference-instrumentation-anthropic",
    "AnthropicInstrumentor",
  ],
  [
    "@arizeai/openinference-instrumentation-langchain",
    "LangChainInstrumentor",
  ],
  [
    "@opentelemetry/instrumentation-openai",
    "OpenAIInstrumentation",
  ],
];

/**
 * Try to instrument all known LLM providers.
 * @returns List of activated instrumentor class names.
 */
export function activateInstrumentors(): string[] {
  const activated: string[] = [];

  for (const [modulePath, className] of KNOWN_INSTRUMENTORS) {
    try {
      // eslint-disable-next-line @typescript-eslint/no-require-imports
      const mod = require(modulePath) as Record<string, unknown>;
      const Cls = mod[className] as { new (): { instrument(): void } } | undefined;
      if (Cls) {
        new Cls().instrument();
        activated.push(className);
      }
    } catch {
      // Package not installed or failed to activate — skip silently
    }
  }

  if (activated.length === 0) {
    console.debug(
      "pixie: No LLM instrumentors activated. Install provider packages " +
        "(e.g. @arizeai/openinference-instrumentation-openai) for auto-capture."
    );
  }

  return activated;
}
